src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,199 @@
+        return getNode(index).data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.data;
+        node.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeToRemove = getNode(index);
+        if (nodeToRemove == first) {
+            removeFirst();
+        } else if (nodeToRemove == last) {
+            removeLast();
+        } else {
+            nodeToRemove.prev.next = nodeToRemove.next;
+            nodeToRemove.next.prev = nodeToRemove.prev;
+            size--;
+        }
+        return nodeToRemove.data;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        while (current != null) {
+            if ((current.data == null && object == null) || (current.data != null
+                    && current.data.equals(object))) {
+                if (current == first) {
+                    removeFirst();
+                } else if (current == last) {
+                    removeLast();
+                } else {
+                    current.prev.next = current.next;
+                    current.next.prev = current.prev;
+                    size--;
+                }
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        Node<T> current = first;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private void addFirst(T value) {
+        Node<T> newNode = new Node<>(null, value, first);
+        if (first == null) {
+            last = newNode;
+        } else {
+            first.prev = newNode;
+        }
+        first = newNode;
+        size++;
+    }
+
+    private void removeFirst() {
+        if (isEmpty()) {
+            throw new RuntimeException("List is empty");
+        }
+        first = first.next;
+        if (first == null) {
+            last = null;
+        } else {
+            first.prev = null;
+        }
+        size--;
+    }
+
+    private void removeLast() {
+        if (isEmpty()) {
+            throw new RuntimeException("List is empty");
+        }
```

Comment:
don't throw any other exceptions than index out of bounds

Reply:
not fixed